sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala

```diff
@@ -190,20 +190,25 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
       .agg(sum("num"))
       .as[(String, Long)]
 
-    testStream(result, Update)(
-      AddData(inputData, "a" -> 1),
-      CheckLastBatch("a" -> 1L),
-      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(1L, 1L)),
-      AddData(inputData, "a" -> 1), // Dropped
-      CheckLastBatch(),
-      assertNumStateRows(total = Seq(1L, 1L), updated = Seq(0L, 0L)),
-      AddData(inputData, "a" -> 2),
-      CheckLastBatch("a" -> 3L),
-      assertNumStateRows(total = Seq(1L, 2L), updated = Seq(1L, 1L)),
-      AddData(inputData, "b" -> 1),
-      CheckLastBatch("b" -> 1L),
-      assertNumStateRows(total = Seq(2L, 3L), updated = Seq(1L, 1L))
-    )
+    // As of [SPARK-40940], multiple state operator with Complete mode is disabled by default
```

Comment:
hmm, the above scenario dropDuplicates -> aggregation was supported before. So I was wrong - dropDuplicates and also inner join with timestamp equality condition can be followed by a stateful operator in any mode - these operators - dropDuplicates and inner equality join  do not delay the output records. I apologize for the randomization, I think the above scenario is important to continue supporting.

Reply:
I see. No problem at all! 

Just to confirm, in the above suite, sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala, there is a query which is flatMapGroupsWithState followed by an agg in complete mode, it also seems to be supported before. But here after the change it throws because there are two stateful ops in complete mode. Should we also allow this case? I'm not sure here because I remember you mentioned flatMapGroupsWithState could change the eventtime so we should disallow... So is the test wrong..?

Reply:
So every stateful operators have its own characteristics...

1. streaming aggregation

- append mode requires event time column to be set in grouping key. It will emit records which watermark passed by the value of event time column. (delaying the output records in above comment)
- For update mode and complete mode, having event time column in grouping key is optional. It only helps to evict the state. (In complete mode this even doesn't happen.) It won't delay the output records.
- For update mode and complete mode, downstream operator(s) followed by streaming aggregation must handle the outputs streaming aggregation produces semantically properly. E.g. update mode will produce outputs multiple times for the same aggregated output, say, correction. complete mode will produce all the historical aggregated outputs.
- I can't imagine the valid case for update mode and complete mode of streaming aggregation to be followed by another stateful operator.

2. deduplication

- The behavior is same among all output modes. (More clearly, it "ignores" the output mode.)
- It won't delay the output records.
- It produces the same output only once.

So, deduplication operator itself does not have any compatibility limitation. If the combination of stateful operator A and deduplication operator should be blocked, it is due to the operator A.

3. stream-stream join

- I cannot reason about the proper behavior for update and complete mode.
- It only accepts append mode now, so good to retain it.

4. flatMapGroupsWithState

- It doesn't support complete mode at all.
- It supports append mode and update mode, although it's purely relying on user function to do the right thing for the output mode, which I believe most of regular end users wouldn't do.
- It loses the event time column on output of the operator.
- It supports processing time semantic, which IMHO we should never allow this to be used with event time semantic in other stateful operators.

That said, maybe we still need a hand-made enumeration of allow/block list for update/complete mode...

Reply:
If we sort this out properly, then we probably could come up with a couple of (or several) golden rules from here e.g. no stateful operator is supported after stateful operator A in some output mode(s). The list of rules has to still be a hand-made one though.

Reply:
I've made a list, let's discuss this later.

- In Complete, Update mode, Aggregations followed by any stateful op are disallowed
- Dedup: don't count, has no effect no matter what stateful op and output mode is.
- stream-stream join: 
  - only allowed in append mode, inner join with equality.
  - Outer join with equality and time-interval join are disallowed.
  - [?] Other than that, don't need to check its compatibility with other stateful ops.
- flatMapGroupsWithState (and mapGroupWithState, also pandas version): 
  - Currently: `MapGroupsWithState` with aggregation is disallowed 
  - Currently: `MapGroupsWithState` only allowed in Update mode
  - [?] After this PR: `MapGroupsWithState` what?
  - Currently: `flatMapGroupsWithState`'s output mode must match query output mode if no aggs -> [keep this behavior] 
  - Currently: `flatMapGroupsWithState` with agg (no matter before or after it) in Update mode is not allowed -> [keep this behavior]
  - Currently: agg followed by `flatMapGroupsWithState` in Append mode is disallowed -> [change this behavior]
  - After this PR: agg followed by `flatMapGroupsWithState` in Append mode is allowed. 
  - After this PR: `flatMapGroupsWithState` followed by any stateful operator is disallowed.

[?] Why Dedup doesn't require event-time col? It should create some kind of state store to do the deduplication, if no watermark are we holding these states throughout the query?

Reply:
Must works: 
- Join on equality chained with agg in Append mode
- Time-interval join & flatmapgroupswithSate followed by stateful op: disallowed
- Just don't count when dedup doesn't have eventtime 
- ? only allow all stateful followed by dedup -> only in Update and Complete mode

Reply:
Eventually, the above boils down to the simple 3 golden rules:
1. `flatMapGroupsWithState`, `MapGroupsWithState` followed by any stateful operator is disallowed. 
2. Stream-stream time interval join followed by any stateful operator is disallowed. Note that this is only allowed in Append mode.
3. Aggregation followed by any stateful operators is disallowed in Complete and Update mode.